src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,335 @@
+use rusqlite::{Connection, OptionalExtension, NO_PARAMS, Row, Savepoint};
+use rusqlite::types::ToSql;
+
+use vm::contracts::Contract;
+use vm::errors::{Error, ErrType, InterpreterResult as Result, IncomparableError};
+use vm::types::{Value, TypeSignature, TupleTypeSignature, AtomTypeIdentifier};
+
+const SQL_FAIL_MESSAGE: &str = "PANIC: SQL Failure in Smart Contract VM.";
+const DESERIALIZE_FAIL_MESSAGE: &str = "PANIC: Failed to deserialize bad database data in Smart Contract VM.";
+
+pub struct ContractDatabaseConnection {
+    conn: Connection
+}
+
+pub struct ContractDatabase <'a> {
+    savepoint: Savepoint<'a>
+}
+
+pub struct SqliteDataMap {
+    map_identifier: i64,
+    key_type: TypeSignature,
+    value_type: TypeSignature
+}
+
+pub trait ContractDatabaseTransacter {
+    fn begin_save_point(&mut self) -> ContractDatabase<'_>;
+}
+
+impl ContractDatabaseConnection {
+    pub fn initialize(filename: &str) -> Result<ContractDatabaseConnection> {
+        let mut contract_db = ContractDatabaseConnection::inner_open(filename)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS maps_table
+                      (map_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       map_name TEXT,
```

Comment:
Will `map_name` ever be `NULL`?  If not, then this should be `TEXT NOT NULL`.

Reply:
No, it'll never be NULL, so I'll make these fields NOT NULL. I want to rethink the schema of these tables anyhow, so what I'll do is start an issue for smart contracting data schemas.